Add support for contrastive search

skavulya opened the pull request:
# What does this PR do?

Adds support for contrastive search for static and dynamic inputs, and low memory configs. Also adds support for `GPT2DoubleHeadsModel`.

Fixes the tests below:
```
GAUDI2_CI=1  RUN_SLOW=1 python -m pytest  tests/transformers/tests/models/t5/test_modeling_t5.py::T5ModelIntegrationTests::test_contrastive_search_t5 -s -v

GAUDI2_CI=1  RUN_SLOW=1 python -m pytest  tests/transformers/tests/models/gpt2/test_modeling_gpt2.py -s -v -k test_contrastive_search_gpt2

GAUDI2_CI=1  RUN_SLOW=1 python -m pytest tests/transformers/tests/models/gpt2/test_modeling_gpt2.py -s -v -k test_batch_generation_2heads

GAUDI2_CI=1  RUN_SLOW=1 python -m pytest tests/transformers/tests/models/gpt2/test_modeling_gpt2.py -s -v -k test_contrastive_generate_low_memory

GAUDI2_CI=1  RUN_SLOW=1 python -m pytest tests/transformers/tests/models/gpt2/test_modeling_gpt2.py -s -v -k test_contrastive_generate_dynamic_shapes

```

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

skavulya commented on optimum/habana/transformers/generation/utils.py:
```diff
@@ -1239,8 +1243,426 @@ def contrastive_search(
+                    top_k_ids[:, selected_idx].view(-1, 1), **model_kwargs
+                )
+                if token_idx is not None:
+                    # Reset inputs to top_k_ids for static shapes to override selection by token_idx
```
@regisss @jiminha The shape of `top_k_ids` is different from the search input so `prepare_inputs_for_generation` for static shapes returns zeros because `token_idx` is greater than the dimensions of  `top_k_ids`. 

Alternative ways to fix this would be to temporarily set  `token_idx=1` before preparing inputs, or to create a padded  `top_k_ids` for max_length which we can update. Please let me know what approach to use.

jiminha replied:
@libinta  @ssarkar2  could you comment on Soila's question above and review the code?

libinta replied:
@skavulya  we need to pad the top_k_ids for static shape

skavulya replied:
@libinta  I padded the top_k_ids

skavulya commented on optimum/habana/transformers/generation/utils.py:
```diff
@@ -1239,8 +1243,426 @@ def contrastive_search(
+        if not bucket_internal:
+            if bucket_size >= 0:
+                inc = iter(incrementor(bucket_size, prompt_len))
+            if bucket_size > 0:  # TODO: Why is this needed given above statement?
```
@regisss  @jiminha I copied this code from `greedy_search`. I would like clarification on why it is needed. I added the todo comment to confirm whether to keep the statement. 

Should I also modify `is_greedy_or_beam_and_bucket` to include contrastive_search? https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/generation/utils.py#L587

ssarkar2 replied:
When I originally added the assert in greedy, I was seeing that position_ids was not present when I had bucketing. I think I had that assert just in case somehow position ids was passed, and I wanted that assert to be hit, so that we could study what was teh use case and deal with it appropriately.



Yes u can modify is_greedy_or_beam_and_bucket

Btw what are the commands you are testing?

skavulya replied:
@ssarkar2 Thanks for the clarification. I will keep the assertion so that we can study it if we encounter it. I'll add contrastive search to is_greedy_or_beam_and_bucket criteria. 

I have included the pytests that are fixed by this PR in the PR description. They do not test the bucketing feature. Is there a test or script I can run to test it?

skavulya replied:
@ssarkar2  I added the contrastive search arguments to the `run_generation.py` and `README`. 

Here are sample outputs.

Greedy search:
`python run_generation.py --model_name_or_path meta-llama/Llama-2-7b-hf --use_hpu_graphs --use_kv_cache --batch_size 1 --bf16 --max_new_tokens 512 --prompt "DeepMind Company is"`

'DeepMind Company is a British artificial intelligence company. It was founded in 2010 by Demis Hassabis, Mustafa Suleyman, and Shane Legg. The company is based in London, England.\nDeepMind is a subsidiary of Google. In 2014, Google acquired DeepMind for $400 million.\nDeepMind is best known for its work in artificial intelligence. The company has developed a number of AI systems, including AlphaGo, which defeated a professional Go player in 2016.\nDeepMind is also working on a number of other projects, including a system that can help doctors diagnose diseases and a system that can help people learn new skills.\nDeepMind is a British artificial intelligence company that was founded in 2010. The company is best known for its work in developing artificial intelligence systems that can play games and solve problems. DeepMind has also developed a number of other AI systems, including a system that can help doctors diagnose diseases and a system that can help people learn new skills.\nDeepMind is a British artificial intelligence company that was founded in 2010. The company is best known for its work in developing artificial intelligence systems that can play games and solve problems. DeepMind has also developed a number of other AI systems, including a system that can help doctors diagnose diseases and a system that can help people learn new skills. DeepMind is a subsidiary of Google.\nDeepMind is a British artificial intelligence company that was founded in 2010. The company is best known for its work in developing artificial intelligence systems that can play games and solve problems. DeepMind has also developed a number of other AI systems, including a system that can help doctors diagnose diseases and a system that can help people learn new skills. DeepMind is a subsidiary of Google. DeepMind is a British artificial intelligence company that was founded in 2010. The company is best known for its work in developing artificial intelligence systems that can play games and solve problems. DeepMind has also developed a number of other AI systems, including a system that can help doctors diagnose diseases and a system that can help people learn new skills. DeepMind is a subsidiary of Google.\nDeepMind is a

`python run_generation.py --model_name_or_path meta-llama/Llama-2-7b-hf --use_hpu_graphs --use_kv_cache --batch_size 1 --bf16 --max_new_tokens 512 --prompt "DeepMind Company is" --top_k 4 --penalty_alpha 0.6`

DeepMind Company is a British artificial intelligence (AI) company.\nDeepMind Technologies Limited is a subsidiary of Alphabet Inc. that specializes in AI research and its applications. The company's headquarters are in London, United Kingdom.\nIn 2010, Demis Hassabis, Shane Legg, Mustafa Suleyman, and Wojciech Zaremba founded DeepMind, an Artificial Intelligence company. They were joined by Peter Wurman, Othman Larouz, James Kvamqvist, Max Jaderberg, Oriol Vinyals, Quoc Le Nguyen, Ilya Sutskevichen, Sven Scheurell, David Reeves-Crookshank, Tom Schaeffer, Richard Suttner, and other researchers.\nOn 14 May 2014, Google announced that it had acquired DeepMind for $400 million in cash and $400 million in Google equity. The acquisition was completed on 1 August 2014.\nIn 2015, DeepMind's researchers published a paper in Nature that described an algorithm that could learn to play Atari games without being programmed to do so. This was the first time that a computer had been able to master a video game by learning from experience, rather than relying on pre-programmed rules.\nIn 2016, DeepMind's researchers published a paper in Nature Neuroscience that described a neural network that could predict the movements of people with Parkinson's disease with 96% accuracy. This was the first time that a machine had been able to accurately predict the movements of people with a neurological disorder.\nIn 2017, DeepMind's researchers published a paper in Nature Communications that described a neural network that could generate realistic-looking images of human faces. This was the first time that a machine had been able to generate convincing images of human faces.\nIn 2018, DeepMind's researchers published a paper in Nature Machine Intelligence that described a neural network that could solve Rubik's Cube puzzles in 1.2 seconds, which is faster than any human can do it.

libinta commented on optimum/habana/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -571,3 +577,131 @@ def forward(
         )
+
+
+class GaudiGPT2DoubleHeadsModel(GPT2DoubleHeadsModel):
```
please move this to https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/models/gpt2/modeling_gpt2.py and check how to overwrite functions in transformers with git blame to find PR.

libinta commented on optimum/habana/transformers/generation/utils.py:
````diff
@@ -1239,8 +1243,426 @@ def contrastive_search(
         >>> tokenizer.batch_decode(outputs, skip_special_tokens=True)
         ['DeepMind Company is a company that focuses on the development and commercialization of artificial intelligence (AI). DeepMind’s mission is to help people understand and solve problems that are difficult to solve in the world today.\n\nIn this post, we talk about the benefits of deep learning in business and how it']
         ```"""
+        # init values
````
please specify the following code source link + what are the changes.

skavulya replied:
@libinta I've added the code source link and changes

libinta commented on optimum/habana/transformers/generation/utils.py:
```diff
@@ -1239,8 +1243,426 @@ def contrastive_search(
+        logits_processor = logits_processor if logits_processor is not None else LogitsProcessorList()
+        logits_warper = logits_warper if logits_warper is not None else LogitsProcessorList()
+        stopping_criteria = stopping_criteria if stopping_criteria is not None else StoppingCriteriaList()
+        pad_token_id = pad_token_id if pad_token_id is not None else self.generation_config.pad_token_id
```
i don't see https://github.com/huggingface/transformers/blob/v4.40.2/src/transformers/generation/utils.py#L1960 here

skavulya replied:
Yes, I'm using 4.38.2. Would you like me to upgrade the implementation to use 4.40.2?

libinta commented on optimum/habana/transformers/generation/utils.py:
```diff
@@ -1239,8 +1243,426 @@ def contrastive_search(
+
+            model_kwargs["lazy_mode"] = lazy_mode
+
+            # if the first step in the loop, encode all the prefix and obtain: (1) past_key_values;
```
why don't we need prepare_inputs_for_generation for 1st step?

skavulya commented on optimum/habana/transformers/generation/utils.py:
```diff
@@ -1239,8 +1243,426 @@ def contrastive_search(
+            if model_kwargs.get("past_key_values") is None:
+                # prepare inputs
+                model_kwargs["use_cache"] = True
+                model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
```
@libinta prepare_inputs_for_generation  for the first step is called here.

ssarkar2 reviewed (commented):
Can you add an example using run_generation.py with same config of input/output, but one with greedy and one with llama, and post the perf and the sentences generated.

just want to document that perf difference between greedy and contrastive, and make sure teh sentence quality is different from greedy

skavulya commented:
> Can you add an example using run_generation.py with same config of input/output, but one with greedy and one with llama, and post the perf and the sentences generated.
> 
> just want to document that perf difference between greedy and contrastive, and make sure teh sentence quality is different from greedy

@ssarkar2 I'll add the contrastive search arguments to `run_generation.py`. Would you like me to add the examples of the generated text to the README or paste the output here in the comments?

ssarkar2 commented on optimum/habana/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -571,3 +577,131 @@ def forward(
         )
+
+
+class GaudiGPT2DoubleHeadsModel(GPT2DoubleHeadsModel):
```
just curious: this class is added because its needed for contrastive search?

skavulya replied:
I added it to fix the pytest for contrastive search with low memory.

`GAUDI2_CI=1  RUN_SLOW=1 python -m pytest tests/transformers/tests/models/gpt2/test_modeling_gpt2.py -s -v -k test_contrastive_generate_low_memory`

ssarkar2 commented:
> > Can you add an example using run_generation.py with same config of input/output, but one with greedy and one with llama, and post the perf and the sentences generated.
> > just want to document that perf difference between greedy and contrastive, and make sure teh sentence quality is different from greedy
> 
> @ssarkar2 I'll add the contrastive search arguments to `run_generation.py`. Would you like me to add the examples of the generated text to the README or paste the output here in the comments?

I think you just posted a sample output in the discussion here, that should be good enough for documenting it.

skavulya commented:
@ssarkar2 @libinta @jiminha @regisss I have updated the code to use transformers v4.40. Please review and let me know if there are additional changes you would like me to make.

libinta commented:
@skavulya  can you rebase?

skavulya commented:
@libinta @ssarkar2 I rebased the PR. I also updated contrastive search to work with bucketing.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-habana/pr_943). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

regisss commented:
@skavulya When running `make style`, `ruff` found 2 errors:
```
optimum/habana/transformers/generation/utils.py:1958:24: F821 Undefined name `prev_idx`
     |
1956 |                 if model_kwargs.get("token_idx_cpu") <= (model_kwargs["kv_cache_len"] // bucket_size) * bucket_size:
1957 |                     idx = (model_kwargs.get("token_idx_cpu") - 1) // bucket_size
1958 |                     if prev_idx != idx:
     |                        ^^^^^^^^ F821
1959 |                         model_kwargs["cache_idx"] = (idx + 1) * bucket_size
1960 |                         prev_idx = idx
     |

optimum/habana/transformers/generation/utils.py:1960:25: F841 Local variable `prev_idx` is assigned to but never used
     |
1958 |                     if prev_idx != idx:
1959 |                         model_kwargs["cache_idx"] = (idx + 1) * bucket_size
1960 |                         prev_idx = idx
     |                         ^^^^^^^^ F841
1961 |                 else:
1962 |                     model_kwargs["cache_idx"] = model_kwargs["kv_cache_len"]
```
I'm not sure exactly how to fix them, what is the purpose of `prev_idx`?

skavulya commented:
Thanks @regisss I fixed the issue with prev_idx. It is used for bucketing.

regisss reviewed (commented):
LGTM!
Can you also add a test in https://github.com/huggingface/optimum-habana/blob/main/tests/test_text_generation_example.py? It can just be a simple test with GPT2 using `top_k` and `penalty_alpha`.

skavulya commented:
Thanks @regisss  I added the test for contrastive search. 

@regisss @ssarkar2 When adding the test, I noticed that the throughput is really low compared to A100 for gpt2-xl top_k=4, penalty_alpha=0.5: A100: 154.84	tokens/sec vs. Gaudi2: 25.05 tokens/sec. I have narrowed down the slowest code to the following and need guidance on how to optimize these sections of code:

https://github.com/skavulya/optimum-habana/blob/48f925061fff14ac3b6b2707fe9c782122f7e000/optimum/habana/transformers/generation/utils.py#L1852-L1855 

```
next_decoder_hidden_states = ()
for layer in full_hidden_states:
    layer = torch.stack(torch.split(layer, top_k))[range(batch_size), selected_idx, :]
    next_decoder_hidden_states += (layer,)
```

There is a similar section of code here:
https://github.com/skavulya/optimum-habana/blob/48f925061fff14ac3b6b2707fe9c782122f7e000/optimum/habana/transformers/generation/utils.py#L1877-L1925

yeonsily commented:
@skavulya can you please fix code style?

skavulya commented:
@yeonsily I fixed the code style. 

@regisss @ssarkar2 I improved the performance of the code by creating the range indices on hpu instead of gpu. The performance of gpt2 improved from 98 tokens/sec to 173 tokens/sec for the following command:

`python run_generation.py --model_name_or_path gpt2 --use_hpu_graphs --use_kv_cache --max_new_tokens 100 --top_k 4 --penalty_alpha 0.5 --prompt "Here is my prompt"`

skavulya commented:
@regisss I saw the PR for upgrade to 4.43 https://github.com/huggingface/optimum-habana/pull/1163 Would you like me to upgrade this PR too?

regisss commented:
> @regisss I saw the PR for upgrade to 4.43 #1163 Would you like me to upgrade this PR too?

That would be great if you have time to do it!
Otherwise I'll try to do it in the next few days.

regisss reviewed (approved):
LGTM!

I pushed new changes to align it with Transformers 4.43 and added a baseline for Gaudi1.

skavulya commented:
Thanks @regisss!
